File: haskell/glue/src/Glue/Lib/Math/Utility/Max.hs
module Glue.Lib.Math.Utility.Max where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Maximum function
-- Mirrors Haskell Glue.Lib.Math.Utility.Max.max exactly
max :: IR Eval
max = NativeFunc maxImpl

-- Maximum function implementation
-- Mirrors Haskell Glue.Lib.Math.Utility.Max.maxImpl exactly
maxImpl :: IR Eval -> Eval (IR Eval)
maxImpl arg1 = pure $ NativeFunc (maxWith arg1)

maxWith :: IR Eval -> IR Eval -> Eval (IR Eval)
maxWith arg1 arg2 = case (arg1, arg2) of
    (Integer n1, Integer n2) -> pure $ Integer (Prelude.max n1 n2)
    (Float n1, Float n2) -> pure $ Float (Prelude.max n1 n2)
    (Integer n1, Float n2) -> pure $ Float (Prelude.max (fromIntegral n1) n2)
    (Float n1, Integer n2) -> pure $ Float (Prelude.max n1 (fromIntegral n2))
    _ -> throwError $ wrongArgumentType ["number", "number"]

File: dart/glue/lib/src/lib/math/utility/max.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Maximum function (returns the larger of two numbers)
/// Mirrors Haskell Glue.Lib.Math.Utility.Max.max exactly
final Ir max = IrNativeFunc(maxImpl);

/// Maximum function implementation (returns the larger of two numbers)
/// Mirrors Haskell Glue.Lib.Math.Utility.Max.maxImpl exactly
Eval<Ir> maxImpl(Ir left) {
  return Eval.pure(IrNativeFunc(maxWith(left)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Math.Utility.Max.maxWith exactly
Eval<Ir> Function(Ir) maxWith(Ir left) {
  return (Ir right) {
    return sequenceAll([eval(left), eval(right)]).flatMap((values) {
      return switch (values) {
        [IrInteger(value: final n1), IrInteger(value: final n2)] => Eval.pure(
          IrInteger(n1 > n2 ? n1 : n2),
        ),
        [IrFloat(value: final n1), IrFloat(value: final n2)] => Eval.pure(
          IrFloat(n1 > n2 ? n1 : n2),
        ),
        [IrInteger(value: final n1), IrFloat(value: final n2)] => Eval.pure(
          IrFloat(n1 > n2 ? n1.toDouble() : n2),
        ),
        [IrFloat(value: final n1), IrInteger(value: final n2)] => Eval.pure(
          IrFloat(n1 > n2 ? n1 : n2.toDouble()),
        ),
        _ => throwError(wrongArgumentType(['number', 'number'])),
      };
    });
  };
}
